### DISCUSSION TITLE
Cloud encryption and data storage

### DISCUSSION THREAD

**Discussion by mockinbirdy** [toeverything/AFFiNE#6160] [General]
First off, this FAQ response sounds like a company trying to be clever with their words and not directly answering the question. Using overly descriptive language instead of simply saying "yes" or "no" and giving a full explanation of how it works in their app. To make matters worse, it only mentions local network and doesn't say a word about cloud storage.

> Q: Does AFFiNE provide support for end-to-end encryption?
> 
> A: Certainly, AFFiNE extends robust support for end-to-end encryption (E2EE). Our platform prioritizes data security while fostering seamless teamwork within local networks. With AFFiNE, you can confidently leverage E2EE alongside local area network (LAN) collaboration. Moreover, our collaboration framework, based on Conflict-free Replicated Data Types (CRDTs), ensures that real-time collaboration remains unhindered by geographical constraints, network speeds, and even offline usage. We anticipate these improvements to be rolled out by the end of this year.

So... 

How does end-to-end encryption work in AFFiNE cloud sync? 
How is user data stored on AFFiNE servers? 

Given that it's known that AFFiNE collects telemetry, it's important to clarify these points.

It's not explained in the FAQ, and there's no way to test it. This raises some red flags. It seems odd that a company promoting itself as an open-source Notion alternative with encryption would not proudly display such a significant feature with a clear and concise explanation, instead of being vague about it.

**Comment by EnterMan123** [comment]
That's one of my concerns

**Reply by ezxpro** [reply]
Came here looking precisely for that answer, since this was posted on march and now it's august and no one answered, I'll assume they just don't want to answer, which makes me think there's some shady business going on with this app's privacy.

But I still that's not the case, I'd really like some of the devs to take this discussion over @forehalo

**Comment by linkh2010** [comment]
I'd like to know how the data is stored in the database as well.

**Comment by forehalo** [comment]
I understand your concerns regarding privacy and data security. First and foremost, I assure you that users’ data is safe, and AFFiNE always prioritizes the privacy and security of its customers’ information.

During data transmission, AFFiNE uses secure protocols to prevent any potential interception, eavesdropping, or tampering. User data is stored in the database as unreadable binary code, ensuring it is not accessible in plain text. AFFiNE is continuously committed to enhancing the secure storage of user data.

Regarding telemetry, I would like to clarify that all data collected by AFFiNE is **anonymous**. This only includes functionality usages, program errors, and crash data, which are used solely to help us improve AFFiNE. Additionally, AFFiNE offers an option to completely disable telemetry or even run AFFiNE client in 100% offline(without sending any network request), and this setting is prominently available for users in the application settings.

Moreover, all of AFFiNE’s client-side code and release processes are openly available on GitHub, allowing the community to review and audit every action we take.

**Reply by negue** [reply]
> AFFiNE uses secure protocols to prevent any potential interception, eavesdropping, or tampering

are these protocols E2EE? or "just" HTTPS?

**Reply by luisliz** [reply]
I hate when companies say this: "During data transmission, AFFiNE uses secure protocols to prevent any potential interception, eavesdropping, or tampering.". Especially to technical people. I have https in my local servers, this is not special. Just like when VPNs say it's "military grade encryption". 

I self-hosted for this reason and because I couldn't trust the AI (just by the nature of how AI works it can't be E2EE) and if they did do AI with E2EE they would say it every chance they could like notesnook does.

**Reply by working-name** [reply]
> Regarding telemetry, I would like to clarify that all data collected by AFFiNE is **anonymous**.

You're **severely** misinformed: https://github.com/toeverything/AFFiNE/issues/12465#issuecomment-3263957706
All data collected is in fact **_NOT anonymous_** at all. It's transmitting my email and name to https://telemetry.affine.run.

**Comment by forehalo** [comment]
I would also like to add a clarification regarding the storage location of user data:

For users accessing app.affine.pro via a web browser:

• All **local workspace** data is stored within the browser itself, and AFFiNE has no way to access it.
• All **cloud workspace** data is stored on AFFiNE’s servers, following the secure storage methods described in my previous response.

For users of the AFFiNE Desktop client:

• All **local workspace** data is stored on the user’s PC file system, and AFFiNE has no way to access it.
• All **cloud workspace** data is stored on AFFiNE’s servers, following the secure storage methods described in my previous response.

For users running a self-hosted instance:

• All **local workspace** data is stored within the browser, and AFFiNE has no way to access it.
• All **cloud workspace** data is stored on the user’s own server, following the secure storage methods described in my previous response, and AFFiNE has no way to access that data.

**Reply by ezxpro** [reply]
With all due respect, your answers are as good as having said nothing. You still didn't adress stuff like:
1. How, exactly, do you treat and store people's data?
2.  What protocols do you use, when, and to what data it applies?
3.  Do you sell user information?
4.  Do you use user data to monetize the platform in any way?  If so, which data is used, how and to what extent? 

That's the sort of stuff users want to know. So far all you've said seems to be far too vague...

**Reply by forehalo** [reply]
No data was applied to do anything. AFFiNE did not ever use any of users' data in any way, for selling, or to monetize, and we will never do that. That's our willing from the very beginning.

we've clarified over and over again. you are so welcomed to audit any actions AFFiNE has ever taken or will take.

**Reply by pstashp** [reply]
First of all, thank you for being responsive and answering our questions.

From browsing the code, I believe the answer to all this back-and-forth is the following (please correct any mistakes):
- No, the database isn't encrypted but in binary format (still considered plaintext, though)
    - I also don't see any KMS code to handle and store the derived encryption keys, so kind of supporting my guess here.
    - Also not seeing any code to handle shared keys for shared content (not sure how that is done in general, though).
- Transport layer is encrypted, but that's a given in today's standard, I guess.

**Reply by ezxpro** [reply]
@pstashp in the end, does this mean that data's not encrypted at rest?

**Reply by ezxpro** [reply]
@forehalo I see you didn't come back to this post. To me it seems like all you said above comes down to "we respect your privacy BUT with no E2EE".

When you say "User data is stored in the database as unreadable binary code" I'm really left wondering what is this even supposed to mean. It's a simple yes or no question. It either is or is not encrypted at rest. 

Also it reads just like @mockinbirdy said;

>  " sounds like a company trying to be clever with their words and not directly answering the question"

You talked and talked and talked here and didn't answer these two simple questions:

> How does end-to-end encryption work in AFFiNE cloud sync?
> How is user data stored on AFFiNE servers?

And no, saying it's in "unreadable binary format" doesn't mean anything. If it's unreadable then how can it be accessed again? Mkes no sense.  All we want to know is this: is it encrypted at rest?
1. YES
2. NO

As simple as that.

**Reply by ezxpro** [reply]
@forehalo still waiting for a comment...

**Reply by ezxpro** [reply]
Just so you know guys, according to support, affine doesn't support cloud encryption:
![image](https://github.com/user-attachments/assets/f092ca99-041a-4148-ab8e-b61a01e0cb40)

Turns out we were right, just as @mockinbirdy (and I too) had suspected. Despite being "clever with their words", they don't support encryption at rest, which is okay.

What is not okay is the extent to which an effort was made not to disclose this, as we've seen @forehalo do.

**Reply by lsadehaan** [reply]
I arrived here looking for a secure knowledge base solution (with at least at-rest encryption, but hoping for E2EE). It's quite amazing how they are trying to claim secure storage but then says "User data is stored in the database as unreadable binary code". This is either an attempt to deceive or a severe lack of competence - not sure which is worse.
Guess my search continues.

**Reply by ezxpro** [reply]
acreom supports e2ee and imo is being open sourced. It's the closest to notion that supports encryption that I could find.
I'm eyeing it as the best candidate to take all my data out of notion, but it still doesn't support file attachments. Once it does I'll say goodbye to notion

**Reply by ovizii** [reply]
> acreom supports e2ee and imo is being open sourced. It's the closest to notion that supports encryption that I could find. 
Never heard about it, so I did a quick search, and it has been open source since June last year, check out their blog.

**Reply by lsadehaan** [reply]
That seems great for a notes app, but I really need document storage - specifically want to use it for very securely storing and sharing only with authorized users documents received under NDA.
I just found a project that seems to have a lot of potential - Peergos. I haven't installed it yet, but from a security perspective it seems to be great and it also stores any files including file preview for PDF at least. It aims to be a distributed social media platform. Gonna try it out.

**Reply by DanielProg39** [reply]
> I arrived here looking for a secure knowledge base solution (with at least at-rest encryption, but hoping for E2EE). It's quite amazing how they are trying to claim secure storage but then says "User data is stored in the database as unreadable binary code". This is either an attempt to deceive or a severe lack of competence - not sure which is worse. Guess my search continues.

There's one option that meets all criteria: [Anytype](https://anytype.io/). It is a full-fledged knowledge base, fully open source, local-first, with end-to-end encrypted P2P sync and cloud backup. The free tier provides 100 MB of E2EE cloud storage.

**Reply by dr3st** [reply]
You can use anytype, but to be honest, the self hosted version is unable to purge unused data to free up space :D
There are some issues / reports, but no changes so far. Could be hard to find out what data is uneeded, when you cannot look into data to check for needance. This might not be an point against anytype, but just in case someone would like to know.

**Comment by mockinbirdy** [comment]
Oh, what a circus. Just the fact that this company couldn’t give a straight yes-or-no answer for 6+ months shows how little they value their customers. Such a smol little start-up, yet it’s already acting like a big corpo with dollar signs in their eyes.

**Comment by Qenjin** [comment]
Main reason I switched from Notion is privacy. These apps are used to create businesses, store sensitive data, and even be used as a personal journal, which is a core feature of AFFiNE. Privacy and security should be the number one feature in my opinion. I'm sure I'm not the only one.

**Comment by luisliz** [comment]
I initially selfhosted the instance precisely to see how the content is stored and if it was encrypted. The data is not easily readable but definitely could be, specially considering you can just go and read the source code. I love this software, but hate the weird avoidance of this question. I basically had to get the answer myself, because the question is never answered straight up, other than the usual "secure protocol", which is just https. 

Just be honest. Technical people who are in github are very likely to be able to read the code anyways and to understand what the word salad really means.

**Reply by dr3st** [reply]
Did the same, you can read at least fragments of data, which can be an issue for secrets ...
Even the request for local encryption (https://github.com/toeverything/AFFiNE/issues/5491) is still "ongoing".
Data encryption (except transport) seems not to be a priority here :/
